_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+
+```scala
+package rockthejvm.websockets
+
+import cats.effect.IOApp
+import cats.effect.IO
+import Server.server
+
+object Program extends IOApp.Simple:
+  override def run: IO[Unit] = server[IO]
+```
+
+Above, we define the `run()` function by calling `server[IO]`. Finally, we can run our application by calling `sbt run` in the terminal. When we navigate to `localhost:8080/chat.html` in our browser we should get the following page:
+
+![Chatbox](../images/websockets/01.png)
+
+## 3. Websocket Introduction
+
+In this section, we'll learn some important information needed to understand to code that follows. WebSockets send and receive data in a data structure called a WebSocket Frame that can be divided into two types, namely, **Data frames** and **Control frames**.
+
+### 3.1. Data Frames
+
+Data Frames are used by WebSockets to carry application-layer and/or extension-layer data. There are two types of data frames, **Text** and **Binary** data frames each identified by an opcode where the most significant bit of the opcode is 0.
+
+The **opcode** determines the interpretation of the data, where 0x1 and 0x2 are opcodes used for Text and Binary data frames respectively while 0x3 - 0x7 are reserved for non control frames yet to be determined.
+
+The Text data frame carries its payload as text data encoded in **UTF-8 format** while the Binary data frame carries its payload as arbitrary **binary data** whose interpretation is solely up to the application layer.
+
+### 3.2. Control Frames
+
+Control frames are used to communicate the **state** of the WebSocket and can be identified by opcodes where the most significant bit of the opcode is 1. These must have a payload length of 125 bytes or less and must not be fragmented.
+
+There are three types of control frames, which include, **Close**, **Ping**, and **Pong** frames whose opcodes are 0x8, 0x9, and 0xA respectively.
+
+A Close frame is sent when the WebSocket is closed or disconnected and may contain a body with the reason for closing. 
+
+The reason may or may not be human readable but can be used for debugging purposes moreover the application must not send any more data frames after sending a Close frame.
+
+A Ping frame may contain application data and when received, the endpoint must respond with a Pong frame containing the same application data that was sent. Ping frames can serve as a means to verify that the endpoint is still responsive creating a heartbeat of the WebSocket.
+
+## 4. Websocket Implementation
+
+To implement the WebSocket in Http4s, we'll need to add another route in `Routes.scala` as follows:
+
+```scala
+import org.http4s.server.websocket.WebSocketBuilder2
+import fs2.Stream
+import org.http4s.websocket.WebSocketFrame
+import fs2.Pipe
+
+object Routes{
+  def service[F[_]: Files: MonadThrow](
+    wsb: WebSocketBuilder2[F]
+  ): HttpApp[F] = {
+    val dsl = new Http4sDsl[F] {}
```

Comment:
make Routes extend Http2sDsl rather than build it here

Reply:
Fixed